glue/core/roi.py

```diff
@@ -614,11 +600,57 @@ def __init__(self, roi_2d=None, projection_matrix=None):
         self.projection_matrix = np.asarray(projection_matrix)
 
     def contains3d(self, x, y, z):
-        """"Test if the projected coordinates are contained in the 2d roi."""
+        """
+        Test whether the projected coordinates are contained in the 2d ROI.
+        """
+
         if not self.defined():
             raise UndefinedROI
-        x, y = _project(self.projection_matrix, x, y, z)
-        return self.roi_2d.contains(x, y)
+
+        x = np.asarray(x)
+        y = np.asarray(y)
+        z = np.asarray(z)
+
+        # Since the projection can significantly increase the memory usage, we
+        # do the following operation in chunks.
+
+        original_shape = x.shape
+
+        # FIXME: this can cause a copy to be made if the data is not
+        # contiguous (for example for broadcasted pixel coordinates).
+        x = x.ravel()
+        y = y.ravel()
+        z = z.ravel()
+
+        chunk_size = 1000000
+        array_size = len(x)
+
+        n_chunks = max(array_size // chunk_size, 1)
```

Comment:
I think this will fail for an array_size of 2000001, since that should be 3 chunks. `int(math.ceil(array_size / chunk_size))` should work (without the max).

Reply:
Thanks! I've now refactored the chunking to use dedicated functions which operate a little differently (basically I am trying to avoid having to change the arrays to 1D)